### PR TITLE
Improve cbauth integration

### PR DESCRIPTION
**Pull request by brett19**

### CODE REVIEW

File: cbauthx/cbauth.go [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	password string,
+	clusterUuid string,
+) (*CbAuthClient, string, error) {
+	// Do not attempt to connect to cbauth faster than allowed
+	{
+		lastConnectDelta := time.Since(a.lastConnectTime)
+		if lastConnectDelta < a.minReconnectTime {
+			select {
+			case <-time.After(a.minReconnectTime - lastConnectDelta):
+			case <-ctx.Done():
```
Comment (chvck) [inline cbauthx/cbauth.go:165]: Should this do something other than just keep running?

File: cbauthx/cbauth.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+		a.lastConnectTime = time.Now()
+	}
+
+	var availableEndpoints []string
+	var failedEndpoints []string
+	var failedErrors []error
+	for {
+		availableEndpoints = availableEndpoints[:0]
+
+		a.lock.Lock()
```
Comment (chvck) [inline cbauthx/cbauth.go:178]: Why is there a new to lock here?

File: cbauthx/cbauth.go [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+			lock.Unlock()
+
+			if a.closed {
+				break
+			}
+		}
+	}()
+
+	for {
+		if initialCli != nil {
```
Comment (chvck) [inline cbauthx/cbauth.go:296]: Confused what this is doing, surely cli is already assigned to initialCli if initialCli isn't nil?

File: cbauthx/cbauth.go [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+func (a *CbAuth) CheckUserPass(ctx context.Context, username string, password string) (UserInfo, error) {
+	cli := a.currentClient.Load()
+	info, err := cli.CheckUserPass(ctx, username, password)
+	if err != nil {
+		if errors.Is(err, ErrInvalidAuth) {
+			return UserInfo{}, ErrInvalidAuth
+		}
+
+		a.logger.Debug("failed to check user with cbauth", zap.Error(err))
```
Comment (chvck) [inline cbauthx/cbauth.go:374]: Is this technically request level logging?

File: cbauthx/cbauthclient.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	return cli, nil
+}
+
+func (c *CbAuthClient) refreshLivenessLocked() {
+	// store the latest time for the heartbeat
+	c.lastCommTs.Store(time.Now().Unix())
+
+	// Reset our timer so we don't close the connection
+	if c.heartbeatTimer != nil {
+		c.heartbeatTimer.Reset(c.livenessTimeout)
```
Comment (chvck) [inline cbauthx/cbauthclient.go:176]: It's slightly confusing that this is a different timeout to just above.

File: cbauthx/cbauth.go (hunk 1 above)
Comment (brett19) [reply cbauthx/cbauth.go:165]: Yes.  Good point.

File: cbauthx/cbauth.go (hunk 2 above)
Comment (brett19) [reply cbauthx/cbauth.go:178]: A mistake from when I moved the endpoints to being read from inside the function, to being passed in.

File: cbauthx/cbauth.go (hunk 3 above)
Comment (brett19) [reply cbauthx/cbauth.go:296]: This is true, but I wanted the logic to be more consolidated.  I'll adjust the variable declaration to no longer default to initialCli though.

File: cbauthx/cbauth.go (hunk 4 above)
Comment (brett19) [reply cbauthx/cbauth.go:374]: It's at debug level, so should be fine.

File: cbauthx/revrpcclient_test.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package cbauthx
+
+/*
```
Comment (chvck) [inline cbauthx/revrpcclient_test.go:3]: hmm

Comment (brett19) [reply cbauthx/revrpcclient_test.go:3]: Ahh, forgot to remove this from the commit.

**Review by chvck — APPROVED** [review]

**Pull request merged**